overexposed: fix and clean up color profile handling

flannelhead opened the pull request:
The rendering intent of the conversion from source (display) profile to histogram profile was changed from perceptual to relative colorimetric in 532a1c5191ee64f0ba60386b34191dd591885202. However, there was duplicated code doing the same task in the OpenCL path which wasn't changed accordingly.

Also use the same histogram profile info for the luminance calculation inside `process()` for consistency. Remove duplication of code that already exists in `iop_profile.c`.

Instead of hardcoded display profile, use whichever profile is used in the current part of the pixelpipe. This prepares for the possibility to move `overexposed` to an earlier position in the pixelpipe, right before `colorout`.

One possible issue that was noticed so far: with this PR, if OpenCL is not used, with a certain device-specific display profile I couldn't get luminance upper clipping indicator to show at all with the default upper limit. This may be related to differences between CPU and OpenCL implementations in the details of how clipping and LUTs are handled.

flannelhead commented:
Rebased on master to fix conflicts. This patch removes a FIXME comment recently added by @dtorop:

```
-  // if in gamut check use soft proof
-  // FIXME: This isn't actually true -- we only use soft proof profile if it is chosen by the user as a histogram profile. Otherwise, if in gamut check, we still use the histogram profile.
```

The patch doesn't alter this behaviour - i.e. `overexposed` will use selected histogram profile regardless of soft proof mode being selected (see changes in `iop_profile.c`). A good question is whether this is actually desired and intended behaviour - should `overexposed` indeed use the softproof profile when gamut check mode is activated?

TurboGit commented on src/common/iop_profile.c:
```diff
@@ -851,13 +851,29 @@ dt_ioppr_set_pipe_output_profile_info(struct dt_develop_t *dev,
 }
 
-dt_iop_order_iccprofile_info_t *dt_ioppr_get_histogram_profile_info(struct dt_develop_t *dev)
+static void _dt_ioppr_get_histogram_profile_type(dt_colorspaces_color_profile_type_t *profile_type,
```
static internal function should not start with "_dt" just "_ioppr_...". I know there is some wrong naming but they should be eliminated. That's just a convention, right.

flannelhead replied:
Alright, thanks for noticing this! On the other hand, could it just be `_get_histogram_profile_info` (as the `ioppr_` part would also have been a "namespace" but this function is internal)?

TurboGit replied:
Yes, fine by me and probably better indeed as we generally remove this in other places.

TurboGit commented on src/common/iop_profile.c:
```diff
@@ -851,13 +851,29 @@ dt_ioppr_set_pipe_output_profile_info(struct dt_develop_t *dev,
+
+static dt_iop_order_iccprofile_info_t *_dt_ioppr_get_histogram_profile_info(struct dt_develop_t *dev,
+                                                                            const int intent,
+                                                                            const int force_normal_mode)
```
force_normal_mode looks like a boolean value, should be a gboolean

there is similar change to be done in other functions below.

flannelhead replied:
Indeed, thanks.

TurboGit commented on src/common/iop_profile.c:
```diff
@@ -851,13 +851,29 @@ dt_ioppr_set_pipe_output_profile_info(struct dt_develop_t *dev,
+
+dt_iop_order_iccprofile_info_t *dt_ioppr_get_histogram_profile_info(struct dt_develop_t *dev)
+{
+  return _dt_ioppr_get_histogram_profile_info(dev, INTENT_PERCEPTUAL, 0);
```
0 -> FALSE

TurboGit commented on src/common/iop_profile.c:
```diff
@@ -851,13 +851,29 @@ dt_ioppr_set_pipe_output_profile_info(struct dt_develop_t *dev,
+
+dt_iop_order_iccprofile_info_t *dt_ioppr_get_histogram_profile_info_for_overexposed(struct dt_develop_t *dev)
+{
+  return _dt_ioppr_get_histogram_profile_info(dev, INTENT_RELATIVE_COLORIMETRIC, 1);
```
1 -> TRUE

TurboGit commented on src/common/iop_profile.c:
```diff
@@ -997,10 +1013,12 @@ void dt_ioppr_get_export_profile_type(struct dt_develop_t *dev,
-                                         const char **profile_filename)
+static void _dt_ioppr_get_histogram_profile_type(dt_colorspaces_color_profile_type_t *profile_type,
+                                                 const char **profile_filename,
+                                                 const int force_normal_mode)
```
gboolean

TurboGit commented on src/iop/overexposed.c:
```diff
@@ -185,10 +138,17 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
+    dt_ioppr_transform_image_colorspace_rgb(in, img_tmp, roi_out->width, roi_out->height, current_profile,
+                                            work_profile, self->op);
+  else
+    fprintf(stderr, "[_transform_image_colorspace] can't create transform profile\n");
```
inside brackets we put the name of the failing function, here it is process in overexpose so I'll use "overexposed.process".

flannelhead replied:
Checked the convention from other iop files, seems there is a space instead of dot between iop name and function. But anyway, changed. Thanks.

TurboGit commented on src/iop/overexposed.c:
```diff
@@ -467,9 +409,12 @@ int process_cl(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, cl_m
+    dt_ioppr_transform_image_colorspace_rgb_cl(devid, dev_in, dev_tmp, roi_out->width, roi_out->height,
+                                               current_profile, work_profile, self->op);
+  else
+    fprintf(stderr, "[_transform_image_colorspace_cl] can't create transform profile\n");
```
overexposed.process_cl

TurboGit reviewed (changes requested):
Some comments.

If I read this correctly it is mostly a clean-up using routines in iop_profile instead of duplicating the code in overexposed. A welcomed clean-up. Not tested yet.

TurboGit commented:
> with a certain device-specific display profile I couldn't get luminance upper clipping

Can you tell me more ? What device-specific display profile will broke this ?

TurboGit commented:
Oh I see that you have an issue #7660 opened, but it is not referenced here so I couldn't tell. Please add:

Fixes #7660.

In you commit message. TIA.

flannelhead commented:
@TurboGit Thank you for the comments. Should be resolved now.

> If I read this correctly it is mostly a clean-up using routines in iop_profile instead of duplicating the code in overexposed. A welcomed clean-up. Not tested yet.

Mostly a cleanup but also contains (and started as) a bugfix. Namely, the OpenCL path was still using perceptual rendering intent for converting to histogram profile. CPU path was changed to relative colorimetric in https://github.com/darktable-org/darktable/commit/532a1c5191ee64f0ba60386b34191dd591885202

Also, `process` and `process_cl` did a separate call to `dt_ioppr_get_histogram_profile_info()` to get the histogram profile for luminance calculations, and as a result the perceptual rendering intent was applied there. This seemed wrong to me as `dt_ioppr_transform_image_colorspace_rgb()` used the histogram color profile with the perceptual rendering intent. (Although, as Aurélien wrote the new clipping indicator modes, might be good to get a confirmation from him.)

Hence it made sense to remove the various different calls that fetched the histogram profile and consolidate them into a single place that always gave histogram profile with relative colorimetric intent. That's why the bugfixes and cleanups are contained in a single commit. Sorry about that.

> Can you tell me more ? What device-specific display profile will broke this ?

It seems https://github.com/darktable-org/darktable/pull/7652 fixed this for me.

The issue was that when OpenCL was deactivated, histogram profile = linear Rec2020 and the display profile is the one attached in https://github.com/darktable-org/darktable/issues/7660, I couldn't get the clipping indicator to show at all with the upper threshold at the default (99.99 %) despite increasing exposure very high. OpenCL activated the same issue didn't appear. But also without OpenCL, with https://github.com/darktable-org/darktable/pull/7652 applied, I can no longer reproduce this issue. So I guess we're good to go with this one... Although I still wonder why the issue showed up only on the CPU path even though https://github.com/darktable-org/darktable/pull/7652 reversed the matrix multiplication for both CPU and OpenCL paths.

For me, this PR at least improves the situation with some of the nasty behaviour near the black point, that's described in this issue: https://github.com/darktable-org/darktable/issues/3271
After these changes it seems to me that the black (lower) clipping indicator behaves more consistently when switching between different display profiles.

> Oh I see that you have an issue #7660 opened, but it is not referenced here so I couldn't tell. Please add:
> 
> Fixes #7660.
> 
> In you commit message. TIA.

Unfortunately this is a different issue and is not fixed by this PR.

dtorop commented:
@flannelhead This looks to be a change for the better. I like seeing all the duplicated code going away, and that this gives potential for overexposed to move to before colorout. I'm glad that this is receiving some attention.

> The patch doesn't alter this behaviour - i.e. overexposed will use selected histogram profile regardless of soft proof mode being selected (see changes in iop_profile.c). A good question is whether this is actually desired and intended behaviour - should overexposed indeed use the softproof profile when gamut check mode is activated?

I'd suggest that you should remove the `force_normal_mode` flag and always use softproof profile when gamut check mode is activated. This simplifies the code, and I'm unclear if the current behavior is intended or is a bug. I think it would be useful for a user to be able to flip to softproof mode and see overexposed flip in a corresponding way. If this is the wrong call, better to find out early on in the path to v3.6.


For the future, on this comment which goes away in this PR:

https://github.com/darktable-org/darktable/blob/56a44ff605b86372493cbe738f9536d0ffec1c19/src/iop/overexposed.c#L155

The histogram processing already converts the image buffer to histogram profile, so overexposed duplicates that work. There could be slight memory/processing savings if it were possible to cache or daisy-chain this work. Of course if overexposed were to move around in the iop order independently of histogram (or convert with a different intent), both conversions will be necessary.


Question: Should the histogram profile always be applied as relative colorimetric, whether in overexposed or via histogram/waveform scope? It seems to me that the viewer of a histogram doesn't want to see a "nice" fixup of out-of-gamut colors, but to see exactly where colors fall (or don't fall) within the chosen colorspace. This would also simplify a codepath (could remove `dt_ioppr_get_histogram_profile_info_for_overexposed()`) if the softproof behavior is also changed.

I also have questions of when/if rendering intent is used. I believe that for LUT profiles the code will fall back to LCMS -- and be intent-aware. For matrix profiles, they are always relative colorimetric, yes? In which case this may be a moot point for many cases...

Apologies if some of this is out of scope (no pun intended) of the current PR.

flannelhead commented:
> I'd suggest that you should remove the force_normal_mode flag and always use softproof profile when gamut check mode is activated.

Agreed. It would also improve consistency across the different image measurements that use the histogram profile.

> Should the histogram profile always be applied as relative colorimetric, whether in overexposed or via histogram/waveform scope? It seems to me that the viewer of a histogram doesn't want to see a "nice" fixup of out-of-gamut colors, but to see exactly where colors fall (or don't fall) within the chosen colorspace. This would also simplify a codepath (could remove dt_ioppr_get_histogram_profile_info_for_overexposed()) if the softproof behavior is also changed.

To me this also sounds like a good idea. I've added the `force_normal_mode` change and this one in this PR as a separate commit for easier evaluation. Notice that this doesn't change the intent for the actual histogram, since the histogram code explicitly uses perceptual intent here:
https://github.com/darktable-org/darktable/blob/50472ecc5fa673bb24ad1ce19f132fab205815ea/src/libs/histogram.c#L283-L284

However, one should still take some care here since `dt_ioppr_get_histogram_profile_type()` is also used in `rgbcurve` and `colorzones` IOPs, and the colorpicker. Might be good to check if the change of intent is also desired there or not.

> I also have questions of when/if rendering intent is used. I believe that for LUT profiles the code will fall back to LCMS -- and be intent-aware. For matrix profiles, they are always relative colorimetric, yes? In which case this may be a moot point for many cases...

Unfortunately I'm not familiar enough with DT's color management code to give a definitive answer. But if it's true that matrix profiles are always relative colorimetric, it would probably make sense to change the intent explicitly to relative colorimetric in the histogram calculation - then if one changes between a LUT and matrix profile as the histogram profile, that wouldn't cause the intent to switch between relative colorimetric and perceptual (currently for the histogram, the rendering intent is set as perceptual).

An aside: when testing this PR, I suggest applying https://github.com/darktable-org/darktable/pull/7652 since `transform_image_colorspace_rgb` is used for the clipping calculation in `overexposed`.

dtorop commented:
> To me this also sounds like a good idea. I've added the `force_normal_mode` change and this one in this PR as a separate commit for easier evaluation.

Looks good from here. If it's hard to tell if the code in master is a bug or something with an undocumented rationale (I looked back through git history, and there's no rationale in commit messages either). I'd say it's best to just do what makes sense.

> Notice that this doesn't change the intent for the actual histogram, since the histogram code explicitly uses perceptual intent here:
> 
> https://github.com/darktable-org/darktable/blob/50472ecc5fa673bb24ad1ce19f132fab205815ea/src/libs/histogram.c#L283-L284

Good point! Probably well worth changing. If you want to do this via this PR, to keep things consistent and tied to this thread?

> However, one should still take some care here since `dt_ioppr_get_histogram_profile_type()` is also used in `rgbcurve` and `colorzones` IOPs, and the colorpicker. Might be good to check if the change of intent is also desired there or not.

It looks like all other uses of `dt_ioppr_get_histogram_profile_type()` (including in tether) are for colorpicker/histogram uses, so it could be good to be consistent in relative colorimetric. But I'm also interested if someone who has thought this out more carefully would advocate for this...

> if it's true that matrix profiles are always relative colorimetric, it would probably make sense to change the intent explicitly to relative colorimetric in the histogram calculation - then if one changes between a LUT and matrix profile as the histogram profile, that wouldn't cause the intent to switch between relative colorimetric and perceptual (currently for the histogram, the rendering intent is set as perceptual).

Yes, this all sounds right.

flannelhead commented:
> Good point! Probably well worth changing. If you want to do this via this PR, to keep things consistent and tied to this thread?

Yep, added now.

> For the future, on this comment which goes away in this PR:

I preserved that comment about caching the result (although the comment moved to a different place) so it is there for future reference. The caching work itself is probably out of scope for this PR but well worth considering.

dtorop commented on src/common/iop_profile.c:
```diff
@@ -857,7 +857,7 @@ dt_iop_order_iccprofile_info_t *dt_ioppr_get_histogram_profile_info(struct dt_de
   dt_ioppr_get_histogram_profile_type(&histogram_profile_type, &histogram_profile_filename);
   return dt_ioppr_add_profile_info_to_list(dev, histogram_profile_type, histogram_profile_filename,
-                                           DT_INTENT_PERCEPTUAL);
+                                           INTENT_RELATIVE_COLORIMETRIC);
```
Should this be `DT_INTENT_RELATIVE_COLORIMETRIC`?

dtorop replied:
This one is a call into dt code so should use DT_* constant.

flannelhead replied:
Let's change it to use the `DT_` prefix.

dtorop commented on src/develop/pixelpipe_hb.c:
```diff
@@ -890,7 +890,7 @@ static void _pixelpipe_pick_live_samples(const float *const input, const dt_iop_
   // display rgb --> histogram rgb
   if(display_profile && histogram_profile)
-    xform_rgb2rgb = cmsCreateTransform(display_profile, TYPE_RGB_FLT, histogram_profile, TYPE_RGB_FLT, INTENT_PERCEPTUAL, 0);
+    xform_rgb2rgb = cmsCreateTransform(display_profile, TYPE_RGB_FLT, histogram_profile, TYPE_RGB_FLT, INTENT_RELATIVE_COLORIMETRIC, 0);
```
Likewise, `DT_INTENT_RELATIVE_COLORIMETRIC`

dtorop replied:
Second thoughts -- as this is a call into LCMS, should use the LCMS constant.

flannelhead replied:
No change needed here then :)

dtorop commented on src/develop/pixelpipe_hb.c:
```diff
@@ -960,7 +960,7 @@ static void _pixelpipe_pick_primary_colorpicker(dt_develop_t *dev, const float *
   // display rgb --> histogram rgb
   if(display_profile && histogram_profile)
-    xform_rgb2rgb = cmsCreateTransform(display_profile, TYPE_RGB_FLT, histogram_profile, TYPE_RGB_FLT, INTENT_PERCEPTUAL, 0);
+    xform_rgb2rgb = cmsCreateTransform(display_profile, TYPE_RGB_FLT, histogram_profile, TYPE_RGB_FLT, INTENT_RELATIVE_COLORIMETRIC, 0);
```
Likewise, `DT_INTENT_RELATIVE_COLORIMETRIC`

dtorop replied:
Likewise, since a call into LCMS, should use LCMS constant, no change needed.

flannelhead replied:
No change needed here either. :)

dtorop commented on src/iop/overexposed.c:
```diff
@@ -185,10 +138,17 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
+    dt_ioppr_transform_image_colorspace_rgb(in, img_tmp, roi_out->width, roi_out->height, current_profile,
+                                            work_profile, self->op);
+  else
+    fprintf(stderr, "[overexposed process] can't create transform profile\n");
```
This keeps extant behavior, to warn and then do rest of work on uninitialized `img_tmp`. Is it better to:

```c
    dt_iop_copy_image_roi(ovoid, ivoid, ch, roi_in, roi_out, TRUE);
    return;
```

(as above when alloc fails) and warn to dt_control_log()? Could also copy `in` to `img_tmp`, and do the work in display profile -- and warn the user -- but that seems like too much trouble for uncertain results.

flannelhead replied:
Right, copying and emitting an error sounds good. Implemented now.

dtorop reviewed (changes requested):
This looks like a lot of changes of the better in this PR. A few notes based on looking at the overall diff...

I'm not totally clear why there are the `DT_INTENT_*` equivalent to LCMS `INTENT_*`, but seems good to keep to the DT constants as long as they are there.

dtorop commented:
> I'm not totally clear why there are the `DT_INTENT_*` equivalent to LCMS `INTENT_*`

Realized that calls into LCMS should not use DT_* constants, and calls into darktable color code _should_ use the DT_* constants. Commented accordingly above, but not sure how to delete the change requests.

flannelhead commented:
Fixed the rendering intent constant prefixes and added graceful handling of failed color conversion. While I was at it, added control log warnings about failed buffer allocation.

dtorop reviewed (changes requested):
One note about potential memory leak if colorspace work fails.

flannelhead commented on src/iop/overexposed.c:
```diff
@@ -185,10 +139,22 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
+  {
+    fprintf(stderr, "[overexposed process] can't create transform profile\n");
+    dt_iop_copy_image_roi(ovoid, ivoid, ch, roi_in, roi_out, TRUE);
+    dt_control_log(_("module overexposed failed in color conversion"));
```
Good catch. This should be resolved now - added a label / `goto` for cleanup and finalization and also modified the other early exit path to use `goto`.

The first two commits can be squashed if it seems necessary.

flannelhead commented:
Fixed the memory leak now. The first two commits can be squashed if desired.

dtorop commented on src/iop/overexposed.c:
```diff
@@ -415,34 +381,14 @@ void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const
     dt_iop_alpha_copy(ivoid, ovoid, roi_out->width, roi_out->height);
 
-  dt_free_align(img_tmp);
+  if(img_tmp != NULL) dt_free_align(img_tmp);
```
`dt_free_align()` should be a no-op if `img_tmp == NULL`

flannelhead replied:
Yes, indeed. Changed.

dtorop reviewed (commented):
Looks good! One small note, but doesn't seem worth changing anything.

dtorop reviewed (approved):
Looks great...

TurboGit commented:
Thanks, working fine on my side. Good work and thanks @dtorop for the review!
